text/0000-natvis.md

```diff
@@ -0,0 +1,428 @@
+              > strong: {...}
+              > weak: {...}
+              > data: size=0, capacity=1
+                > base: {...}
+                  > hash_builder: {...}
+        ...
+```
+
+The debugger shows the structure of the data, not its meaning. It is useless.
+Even the implementor of `regex` would have a hard time knowing how to decode
+this. In reality, when trying to understand the state of the captures
+variable there are several methods defined for a `Captures` type that paint
+the actual picture in terms of the information a Rust developer would like to
+extract from this variable. In order to meaningfully understand what the `Captures`
+type is truly trying to tell us, it would be very helpful to visualize this data
+differently in the debugger.
+
+What we _want_ is something like this:
+
+```text
+> Variables:
+  > captures: {...}
+    > $1: "SOME_CONSTANT"
+    > $2: "42"
+    > $3: "// some developer comment"
+```
+
+This RFC will describe how to support adding Natvis visualizations which is supported by:
+
+* The Windows Debugger (WinDbg)
+* Visual Studio Debugger.
+
+It should be easy for crate developers to add debugger visualizations to their
+crates.
+
+## Supporting Natvis
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis XML files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+When writing Natvis files for C++, developers write a standalone XML document
+and add it to their project. The build system (such as Visual Studio) knows
+how to package the Natvis file into the debug data (the PDB file) for the
+project, and the debugger knows how to find the Natvis XML in each PDB file.
+
+Developers can add one or more standalone Natvis XML files to their crate.
+The `Cargo.toml` file specifies the name of these Natvis files. This is the
+easiest way to add Natvis support to a project.
+
+The advantage of a standalone XML document is that this process is already
+well-understood by many developers. This will help C++ developers move from
+C++ to Rust. It also avoids any need to modify Rust source code. If a code
+base uses code generation (such as `bindgen` or proc-macros), then standalone
+Natvis XML files would be the only way to provide visualizations for those
+types.
+
+### Standalone Natvis XML files
+
+To provide standalone Natvis XML files, developers create a file with the
+`.natvis` file extension. These Natvis files are then specified in the 
+`Cargo.toml` file via a new manifest key.
+
+As an example, consider a crate with this directory structure:
+
+```text
+/Cargo.toml
+  +-- src
+      +-- main.rs
+      +-- main.natvis
```

Comment:
Is the name important or could this be renamed to foo.natvis?

Reply:
I think using the same basename as the crate's root source file has benefits, because it allows the developer to provide additional NatVis files for each crate or crate binary.

For example:

* `${crate_dir}/src/lib.natvis` - gets compiled into the primary crate for a library crate
* `${crate_dir}/src/main.natvis` - gets compiled into the primary crate for binary (exe) crate
* `${crate_dir}/src/bin/foo.natvis` - gets compiled into the `foo` binary
* `${crate_dir}/tests/big_tests.natvis` - gets compiled into the `big_tests` integration test binary
* etc.

Note that a single linker invocation can receive any number of `/NATVIS:...` parameters.

Reply:
As per the example above, the name of the `.natvis` file can be renamed to the developers choice as long as it has the `.natvis` extension.

Reply:
@sivadeilra are you suggesting that Cargo should be aware of these conventions, or that this mapping can be specified by the programmer?

Reply:
This mapping can be specified by the developer and is not something that would be made aware to Cargo.